File: DSA/Queue/Code/Deque.py
# Deque (Double Ended Queue) in Python

class Deque:
    def __init__(self, max_size=100):
        self.arr = [0] * max_size
        self.front = -1
        self.rear = -1
        self.size = 0
        self.max_size = max_size
    
    def is_empty(self):
        return self.size == 0
    
    def is_full(self):
        return self.size == self.max_size
    
    def insert_front(self, value):
        if self.is_full():
            print("Deque overflow - cannot insert at front")
            return
        if self.is_empty():
            self.front = 0
            self.rear = 0
        else:
            self.front = (self.front - 1 + self.max_size) % self.max_size
        self.arr[self.front] = value
        self.size += 1
        print(f"Successfully inserted at front: {value} (size = {self.size})")
    
    def insert_rear(self, value):
        if self.is_full():
            print("Deque overflow - cannot insert at rear")
            return
        if self.is_empty():
            self.front = 0
            self.rear = 0
        else:
            self.rear = (self.rear + 1) % self.max_size
        self.arr[self.rear] = value
        self.size += 1
        print(f"Successfully inserted at rear: {value} (size = {self.size})")
    
    def delete_front(self):
        if self.is_empty():
            print("Deque underflow - cannot delete from front")
            return -1
        value = self.arr[self.front]
        print(f"Successfully deleted from front: {value} (size = {self.size - 1})")
        
        if self.size == 1:
            self.front = -1
            self.rear = -1
        else:
            self.front = (self.front + 1) % self.max_size
        self.size -= 1
        return value
    
    def delete_rear(self):
        if self.is_empty():
            print("Deque underflow - cannot delete from rear")
            return -1
        value = self.arr[self.rear]
        print(f"Successfully deleted from rear: {value} (size = {self.size - 1})")
        
        if self.size == 1:
            self.front = -1
            self.rear = -1
        else:
            self.rear = (self.rear - 1 + self.max_size) % self.max_size
        self.size -= 1
        return value
    
    def display(self):
        if self.is_empty():
            print("Deque is empty")
            return
        print("Deque: ", end="")
        i = self.front
        for count in range(self.size):
            print(self.arr[i], end=" ")
            i = (i + 1) % self.max_size
        print()

# Test
d = Deque()

print("=== Deque (Double Ended Queue) ===\n")

d.display()

print("\nInserting at rear:")
for i in range(1, 4):
    d.insert_rear(i * 10)

d.display()

print("\nInserting at front:")
d.insert_front(5)
d.insert_front(2)

d.display()

print("\nDeleting from front:")
d.delete_front()

print("\nDeleting from rear:")
d.delete_rear()

d.display()

print("\nComplexity Analysis:")
print("Insert Front: O(1)")
print("Insert Rear: O(1)")
print("Delete Front: O(1)")
print("Delete Rear: O(1)")
print("Space: O(n)")
print("Advantage: Flexible insertion/deletion from both ends")

File: DSA/Queue/Code/Deque.java
// Deque (Double Ended Queue) in Java

public class Deque {
    private int[] arr;
    private int front;
    private int rear;
    private int size;
    private int maxSize;
    
    public Deque(int max_size) {
        arr = new int[max_size];
        front = -1;
        rear = -1;
        size = 0;
        maxSize = max_size;
    }
    
    boolean isEmpty() {
        return size == 0;
    }
    
    boolean isFull() {
        return size == maxSize;
    }
    
    void insertFront(int value) {
        if (isFull()) {
            System.out.println("Deque overflow - cannot insert at front");
            return;
        }
        if (isEmpty()) {
            front = 0;
            rear = 0;
        } else {
            front = (front - 1 + maxSize) % maxSize;
        }
        arr[front] = value;
        size++;
        System.out.println("Successfully inserted at front: " + value + " (size = " + size + ")");
    }
    
    void insertRear(int value) {
        if (isFull()) {
            System.out.println("Deque overflow - cannot insert at rear");
            return;
        }
        if (isEmpty()) {
            front = 0;
            rear = 0;
        } else {
            rear = (rear + 1) % maxSize;
        }
        arr[rear] = value;
        size++;
        System.out.println("Successfully inserted at rear: " + value + " (size = " + size + ")");
    }
    
    int deleteFront() {
        if (isEmpty()) {
            System.out.println("Deque underflow - cannot delete from front");
            return -1;
        }
        int value = arr[front];
        System.out.println("Successfully deleted from front: " + value + " (size = " + (size - 1) + ")");
        
        if (size == 1) {
            front = -1;
            rear = -1;
        } else {
            front = (front + 1) % maxSize;
        }
        size--;
        return value;
    }
    
    int deleteRear() {
        if (isEmpty()) {
            System.out.println("Deque underflow - cannot delete from rear");
            return -1;
        }
        int value = arr[rear];
        System.out.println("Successfully deleted from rear: " + value + " (size = " + (size - 1) + ")");
        
        if (size == 1) {
            front = -1;
            rear = -1;
        } else {
            rear = (rear - 1 + maxSize) % maxSize;
        }
        size--;
        return value;
    }
    
    void display() {
        if (isEmpty()) {
            System.out.println("Deque is empty");
            return;
        }
        System.out.print("Deque: ");
        int i = front;
        for (int count = 0; count < size; count++) {
            System.out.print(arr[i] + " ");
            i = (i + 1) % maxSize;
        }
        System.out.println();
    }
    
    public static void main(String[] args) {
        Deque d = new Deque(100);
        
        System.out.println("=== Deque (Double Ended Queue) ===\n");
        
        d.display();
        
        System.out.println("\nInserting at rear:");
        for (int i = 1; i <= 3; i++) {
            d.insertRear(i * 10);
        }
        
        d.display();
        
        System.out.println("\nInserting at front:");
        d.insertFront(5);
        d.insertFront(2);
        
        d.display();
        
        System.out.println("\nDeleting from front:");
        d.deleteFront();
        
        System.out.println("\nDeleting from rear:");
        d.deleteRear();
        
        d.display();
        
        System.out.println("\nComplexity Analysis:");
        System.out.println("Insert Front: O(1)");
        System.out.println("Insert Rear: O(1)");
        System.out.println("Delete Front: O(1)");
        System.out.println("Delete Rear: O(1)");
        System.out.println("Space: O(n)");
        System.out.println("Advantage: Flexible insertion/deletion from both ends");
    }
}
